coverage/hare_coverage.py

```diff
@@ -0,0 +1,319 @@
+        :return: tuple(stdout, stderr)
+        """
+        logging.debug(f"Executing Command : {command!a}")
+        process: Popen[str] = Popen(command.split(sep=delimiter, maxsplit=max_split),
+                                    stdout=PIPE,
+                                    stderr=PIPE)
+        out, err = process.communicate()
+        if process.returncode:
+            raise Exception(
+                f'Command {command!a} exited with error code {process.returncode}.\n'
+                f'Command output: {err.decode("utf-8")}')
+
+        logging.debug(f'Output:\n{out.decode("utf-8")}')
+
+    def is_exist(self, dir_path: str) -> bool:
+        """
+        Check for the existance of the file or directory.
+        :param dir_path: file or directory path
+        :return: bool
+        """
+        return path.exists(dir_path)
+
+    def is_inside_hare(self) -> None:
+        """
+        Check current working directory is hare home directory or not.
+        if not then raise the exception.
+        :return: None
+        """
+        if getcwd().endswith(f"{sep}hare") is False:
+            logging.error("Not running a script from hare home directory.")
+            raise InvalidDirectoryPath("Not running a script from hare home directory.")
+
+
+class Strategy(ABC):
+
+    """Abstract base class for coverage tool."""
+
+    @abstractmethod
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def check_report(self) -> None:
+        """
+        Validate the report generated
+        :return: None
+        """
+        pass
+
+
+class PythonCoverage(Strategy):
+
+    # This is source to test directory map. It will be used run coverage.
+    # All the newly added source and test should be added here for coverage calculation.
+    source_test_map: Dict[str, str] = {
+        "hax/hax": "hax/test/",
+        "provisioning/miniprov/hare_mp": "provisioning/miniprov/test/"
+    }
+
+    def __init__(self, cov_type: str = "xml", report_dir: str = "hare_coverage_report",
+                 source_test_map: Optional[Dict[str, str]] = None):
+        """
+        Initialize coverage params.
+        :param cov_type: coverage type xml or html
+        :param source_test_map: source module to test module map
+        """
+        self.cov_type: str = "html" if cov_type.lower() == "html" else "xml"
+        if source_test_map is None:
+            self.src_tst_map: Optional[Dict[str, str]] = self.source_test_map
+        else:
+            self.src_tst_map: Optional[Dict[str, str]] = source_test_map

```

Comment:
This if/else can be replaced with `or`:
```suggestion
        self.src_tst_map: Optional[Dict[
            str, str]] = source_test_map or self.source_test_map
```

Reply:
BTW you have both `self.source_test_map` and `self.src_tst_map` in place, while their names seem to be VERY similar and moreover they seem to be logically connected. Moreover, I can't distinguish which field stands for what purpose, so there is a risk to use an incorrect field in the code.

Do you really need to have both of the values stored in `self`? If so, can you make the names to differ more to avoid confusion?

Reply:
Optimized